In [187]:
import pandas as pd
import matplotlib.pyplot as plt
import datetime

In [181]:
def zeroDiagonal(dataTable, dimension):
    ''' Replace Dataframe n x n matrix diagonal values with zero '''
    for i in range(dimension):
        dataTable.iloc[i,i]=0

    return(dataTable)

## <center> Read User Engagement Dataset

In [2]:
user_engagement = pd.read_csv('takehome_user_engagement.csv')
user_engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

### View  user_engagement DataFrame statistical info - describe() and info()

In [35]:
user_engagement_sorted = user_engagement.sort_values(by = 'time_stamp', ascending = False)
user_engagement_sorted.describe()

user_id   visited
count  207917.000000  207917.0
mean     5913.314197       1.0
std      3394.941674       0.0
min         1.000000       1.0
25%      3087.000000       1.0
50%      5682.000000       1.0
75%      8944.000000       1.0
max     12000.000000       1.0

In [36]:
user_engagement_sorted.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 207917 entries, 70763 to 178140
Data columns (total 3 columns):
time_stamp    207917 non-null object
user_id       207917 non-null int64
visited       207917 non-null int64
dtypes: int64(2), object(1)
memory usage: 6.3+ MB


In [195]:
# groupby user_id calculate sum of all logins per user and remove records with NaN
user_log = user_engagement.groupby('user_id').sum().dropna()
# filter by 3 or more user logins
active = user_log[user_log>=3].dropna()
active.head()

visited
user_id         
2           14.0
10         284.0
20           7.0
33          18.0
42         342.0

## <center> Read Users Dataset

In [5]:
# use latin encoding to read european characters
users = pd.read_csv('takehome_users.csv', encoding = 'latin')

In [6]:
users.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

### View  users DataFrame statistical info - describe() and info()

In [39]:
users.describe()

object_id  last_session_creation_time  opted_in_to_mailing_list  \
count  12000.00000                8.823000e+03              12000.000000   
mean    6000.50000                1.379279e+09                  0.249500   
std     3464.24595                1.953116e+07                  0.432742   
min        1.00000                1.338452e+09                  0.000000   
25%     3000.75000                1.363195e+09                  0.000000   
50%     6000.50000                1.382888e+09                  0.000000   
75%     9000.25000                1.398443e+09                  0.000000   
max    12000.00000                1.402067e+09                  1.000000   

       enabled_for_marketing_drip        org_id  invited_by_user_id  
count                12000.000000  12000.000000         6417.000000  
mean                     0.149333    141.884583         5962.957145  
std                      0.356432    124.056723         3383.761968  
min                      0.000000      0.000000            3.000000  
25%                      0.000000     29.000000         3058.000000  
50%                      0.000000    108.000000         5954.000000  
75%                      0.000000    238.250000         8817.000000  
max                      1.000000    416.000000        11999.000000

In [40]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
object_id                     12000 non-null int64
creation_time                 12000 non-null object
name                          12000 non-null object
email                         12000 non-null object
creation_source               12000 non-null object
last_session_creation_time    8823 non-null float64
opted_in_to_mailing_list      12000 non-null int64
enabled_for_marketing_drip    12000 non-null int64
org_id                        12000 non-null int64
invited_by_user_id            6417 non-null float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [157]:
# group by creation_source and report count for each feature
source = users.groupby('creation_source').count()
source

object_id  creation_time  name  email  \
creation_source                                             
GUEST_INVITE             2163           2163  2163   2163   
ORG_INVITE               4254           4254  4254   4254   
PERSONAL_PROJECTS        2111           2111  2111   2111   
SIGNUP                   2087           2087  2087   2087   
SIGNUP_GOOGLE_AUTH       1385           1385  1385   1385   

                    last_session_creation_time  opted_in_to_mailing_list  \
creation_source                                                            
GUEST_INVITE                              1588                      2163   
ORG_INVITE                                3188                      4254   
PERSONAL_PROJECTS                          764                      2111   
SIGNUP                                    1898                      2087   
SIGNUP_GOOGLE_AUTH                        1385                      1385   

                    enabled_for_marketing_drip  org_id  invited_by_user_id  
creation_source                                                             
GUEST_INVITE                              2163    2163                2163  
ORG_INVITE                                4254    4254                4254  
PERSONAL_PROJECTS                         2111    2111                   0  
SIGNUP                                    2087    2087                   0  
SIGNUP_GOOGLE_AUTH                        1385    1385                   0

### From grouping the data frame by 'creation_source' and collecting the count, most accounts originated from organization invite (4254 users). Followed by guest invite to an organization (2163 users) and personal project accounts (2111 users). Lowest login rate is for personal projects (764 users) which may infer less likely for accounts for personal use will have "adopted user".

***

In [182]:
# Replace Dataframe n x n matrix diagonal values with zero to allow other features correlation amplified
sourceCorr = zeroDiagonal(source.corr(),9)
sourceCorr

object_id  creation_time      name     email  \
object_id                    0.000000       1.000000  1.000000  1.000000   
creation_time                1.000000       0.000000  1.000000  1.000000   
name                         1.000000       1.000000  0.000000  1.000000   
email                        1.000000       1.000000  1.000000  0.000000   
last_session_creation_time   0.851242       0.851242  0.851242  0.851242   
opted_in_to_mailing_list     1.000000       1.000000  1.000000  1.000000   
enabled_for_marketing_drip   1.000000       1.000000  1.000000  1.000000   
org_id                       1.000000       1.000000  1.000000  1.000000   
invited_by_user_id           0.891571       0.891571  0.891571  0.891571   

                            last_session_creation_time  \
object_id                                     0.851242   
creation_time                                 0.851242   
name                                          0.851242   
email                                         0.851242   
last_session_creation_time                    0.000000   
opted_in_to_mailing_list                      0.851242   
enabled_for_marketing_drip                    0.851242   
org_id                                        0.851242   
invited_by_user_id                            0.829084   

                            opted_in_to_mailing_list  \
object_id                                   1.000000   
creation_time                               1.000000   
name                                        1.000000   
email                                       1.000000   
last_session_creation_time                  0.851242   
opted_in_to_mailing_list                    0.000000   
enabled_for_marketing_drip                  1.000000   
org_id                                      1.000000   
invited_by_user_id                          0.891571   

                            enabled_for_marketing_drip    org_id  \
object_id                                     1.000000  1.000000   
creation_time                                 1.000000  1.000000   
name                                          1.000000  1.000000   
email                                         1.000000  1.000000   
last_session_creation_time                    0.851242  0.851242   
opted_in_to_mailing_list                      1.000000  1.000000   
enabled_for_marketing_drip                    0.000000  1.000000   
org_id                                        1.000000  0.000000   
invited_by_user_id                            0.891571  0.891571   

                            invited_by_user_id  
object_id                             0.891571  
creation_time                         0.891571  
name                                  0.891571  
email                                 0.891571  
last_session_creation_time            0.829084  
opted_in_to_mailing_list              0.891571  
enabled_for_marketing_drip            0.891571  
org_id                                0.891571  
invited_by_user_id                    0.000000

#### As illusterated in correlation table above, features are highly correlated.

***

### Overall, without indepth analysis, it appears that the users associated with an organiztion as a guest or regular user are more likely to be "adopted users".